storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -88,85 +540,241 @@ type ReconcileVolume struct {
 
-	// Set Volume instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, pv, r.scheme); err != nil {
-		return reconcile.Result{}, err
+	// Pending volume: can do nothing but wait for stabilization.
+	if volume.Status.Phase == storagev1alpha1.VolumePending {
+		reqLogger.Info("pending volume cannot be finalized: requeue")
+		return delayedRequeue(nil)
 	}
 
-	// Check if this PV already exists
-	found := &corev1.PersistentVolume{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: pv.Name}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new PersistentVolume", "PersistentVolume.Name", pv.Name)
-		err = r.client.Create(context.TODO(), pv)
-		if err != nil {
-			return reconcile.Result{}, err
-		}
+	// Check if a PV is associated to the volume.
+	pv, err := self.getPersistentVolume(ctx, volume)
+	if err != nil {
+		reqLogger.Error(
+			err, "error while looking for backing PersistentVolume: requeue",
+			"PersistentVolume.Name", volume.Name,
+		)
+		return delayedRequeue(err)
+	}
 
-		// PV created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
+	// If we have a backing PV we delete it (the finalizer will keep it alive).
+	if pv != nil && pv.GetDeletionTimestamp().IsZero() {
+		if err := self.client.Delete(ctx, pv); err != nil {
+			reqLogger.Error(
+				err, "cannot delete PersistentVolume: requeue",
+				"PersistentVolume.Name", volume.Name,
+			)
+			return delayedRequeue(err)
+		}
+		reqLogger.Info(
+			"deleting backing PersistentVolume",
+			"PersistentVolume.Name", pv.Name,
+		)
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvDeletion",
+			"backing PersistentVolume deleted",
+		)
+		return requeue(nil)
 	}
 
-	// PV already exists - don't requeue
-	reqLogger.Info("Skip reconcile: PersistentVolume already exists", "PersistentVolume.Name", found.Name)
+	// If we don't have a PV or it's only used by us we can reclaim the storage.
+	if pv == nil || isPersistentVolumeUnused(request, pv) {
+		return self.reclaimStorage(ctx, request, volume, pv)
+	}
 
-	return reconcile.Result{}, nil
+	// PersistentVolume still in use: wait before reclaiming the storage.
+	return delayedRequeue(nil)
 }
 
-func newPersistentVolumeForCR(cr *storagev1alpha1.Volume) *corev1.PersistentVolume {
+// Build a PersistentVolume from a Volume object.
+//
+// Arguments
+//     volume:       a Volume object
+//     storageClass: a StorageClass object
+//     volumeSize:   the volume size
+//
+// Returns
+//     The PersistentVolume representing the given Volume.
+func newPersistentVolume(
+	volume *storagev1alpha1.Volume,
+	storageClass *storagev1.StorageClass,
+	volumeSize resource.Quantity,
+) *corev1.PersistentVolume {
+	// We only support this mode for now.
+	mode := corev1.PersistentVolumeFilesystem
+	fsType := storageClass.Parameters["fsType"]
```

Comment:
What happens when this value is not set?

Reply:
Good point. I moved the check for `fsType` here instead.